src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java

```diff
@@ -229,19 +229,55 @@ private static final boolean isVaList(FunctionDescriptor desc) {
         return false;
     }
 
+    private static boolean isUnsupported(MemoryLayout layout) {
+        if (layout instanceof ValueLayout) {
+            return UnsupportedLayouts.isUnsupported((ValueLayout)layout);
+        } else if (layout instanceof GroupLayout) {
+            GroupLayout gl = (GroupLayout)layout;
+            for (MemoryLayout ml : gl.memberLayouts()) {
+                if (isUnsupported(ml)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static MemoryLayout isUnsupported(FunctionDescriptor desc) {
+        if (isUnsupported(desc.returnLayout().orElse(null))) {
```

Comment:
This subtly relies on the fact that isUnsupported layouts(null) does the right thing... not sure if that's too clever. Maybe just check if there's a return type, and if so check it?